enterprise/internal/codeintel/stores/dbstore/uploads.go

```diff
@@ -896,112 +962,222 @@ func (s *Store) updateUploadRetention(ctx context.Context, protectedIDs, expired
-WITH locked_uploads AS (
-	SELECT u.id
-	FROM lsif_uploads u
-	WHERE u.id in (%s)
-	ORDER BY u.id FOR UPDATE
-),
-reference_counts AS (
-	SELECT
-		p.dump_id,
-		count(*) AS count
-	FROM lsif_packages p
-	JOIN lsif_references r
-	ON
-		p.scheme = r.scheme AND
-		p.name = r.name AND
-		p.version = r.version AND
-		p.dump_id != r.dump_id
-	WHERE p.dump_id IN (SELECT id FROM locked_uploads)
-	GROUP BY p.dump_id
-)
-UPDATE lsif_uploads u
-SET num_references = COALESCE((SELECT rc.count FROM reference_counts rc WHERE rc.dump_id = u.id), 0)
-WHERE u.id IN (SELECT id FROM locked_uploads)
-`
-
-// UpdateDependencyNumReferences increments (or decrements) the number of references for
-// each dependency of the uploads with any of the given identifiers.
-func (s *Store) UpdateDependencyNumReferences(ctx context.Context, ids []int, decrement bool) (err error) {
-	ctx, endObservation := s.operations.updateDependencyNumReferences.With(ctx, &err, observation.Args{LogFields: []log.Field{
+// UpdateReferenceCounts updates the reference counts of uploads indicated by the given identifiers
+// as well as the set of uploads that would be affected by one of the upload's insertion or removal.
+// The behavior of this method is determined by the dependencyUpdateType value.
+//
+//   - Use DependencyReferenceCountUpdateTypeNone to calculate the reference count of each of the given
+//     uploads without considering dependency upload counts.
+//   - Use DependencyReferenceCountUpdateTypeAdd to calculate the reference count of each of the given
+//     uploads while adding one to each direct dependency's reference count.
+//   - Use DependencyReferenceCountUpdateTypeRemove to calculate the reference count of each of the given
+//     uploads while removing one from each direct dependency's reference count.
+//
+// To keep reference counts consistent, this method should be called directly after insertion and directly
+// before deletion of each upload record.
+func (s *Store) UpdateReferenceCounts(ctx context.Context, ids []int, dependencyUpdateType DependencyReferenceCountUpdateType) (err error) {
+	ctx, endObservation := s.operations.updateReferenceCounts.With(ctx, &err, observation.Args{LogFields: []log.Field{
 		log.Int("numIDs", len(ids)),
 		log.String("ids", intsToString(ids)),
-		log.Bool("decrement", decrement),
+		log.Int("dependencyUpdateType", int(dependencyUpdateType)),
 	}})
 	defer endObservation(1, observation.Args{})
 
 	if len(ids) == 0 {
 		return nil
 	}
 
+	tx, err := s.transact(ctx)
+	if err != nil {
+		return err
+	}
+	defer func() { err = tx.Done(err) }()
+
 	idQueries := make([]*sqlf.Query, 0, len(ids))
 	for _, id := range ids {
 		idQueries = append(idQueries, sqlf.Sprintf("%s", id))
 	}
+	idListQuery := sqlf.Join(idQueries, ", ")
+
+	excludeCondition := sqlf.Sprintf("TRUE")
+	if dependencyUpdateType == DependencyReferenceCountUpdateTypeRemove {
+		excludeCondition = sqlf.Sprintf("u.id NOT IN (%s)", idListQuery)
```

Comment:
idea: Recently found this, reduces arguments sent and variety of versions of this SQL query ending up in query insights.

```suggestion
		excludeCondition = sqlf.Sprintf("u.id NOT = ANY (%s)", pq.Array(ids))
```

Reply:
This isn't actually equivalent 🤔. A unit test failed when I committed this change.

Reply:
that scares me, I did this before for other parts of our code 👀 Actually, did you commit _just_ this? I think it needs some more than that, like removing the variables created before so they aren't unused. (And actually, I think it needs to be `!= ANY` thinking about it again..)

Reply:
Wait, I got the syntax way wrong. It's `NOT (val = ANY(list))`, not `val != ANY(list)`.